### PR TITLE
[DRAFT] MODDATAIMP-860 - Create Service to Split Marc files - rev3

### PR DESCRIPTION
**Pull request by carolegodfrey**
## Purpose
https://issues.folio.org/browse/MODDATAIMP-860 

Refine/Revisit using a more robust idea/solution from Taras which splits files using vertex ReadStream/WriteStream
Aligns with current Async processing in Data Import
Can be extended for added-on processing (after the files are split)
More testable code

Issues with previous algorithm for splitting (not covering all cases for split - for example when file does not split into even chunks - 100 record file split into size of 7 records for example)

https://github.com/folio-org/mod-data-import/pull/261
https://github.com/folio-org/mod-data-import/pull/262

## Approach
- Adds FileSplitWriter class which implements WriteStream Interface
- Adds Test Controller Method `http://localhost:8082/data-import/testFileSplit?key=20K.mrc` which accepts file on S3
-- Streams file from S3
 -- Pipes input stream to FileSplitWriter which performs the splitting based on presence of RT byte in the file
 -- FileSplitWriter writers split chunk files locally and then uploads completed chunk files to S3
- Adds Test Controller method `http://localhost:8082/data-import/testFileSplitLocal?key=10K.mrc` which accepts local file and splits files locally

#### TODOS and Open Questions
- [x] Check if AsyncInputStream is valid for convert of InputStream to ReadStream which can be pipedTo FileSplitWriter
(currently pulled code from gist https://gist.github.com/Stwissel/a7f8ce79785afd49eb2ced69b56335de)
- [x] Clean up code 
- [x] Fix code smells (code complexity)
- [x] Add unit tests.
- [x] Logs for timing to understand where time in spent for splitting process
- [x] Need to decide where temp files get written ( ./storage/upload .path setting is where current uploads get written - /usr/verticles/storage/upload) - S3 library uses system temp folder

### CODE REVIEW

**Comment by evaluk** [comment]
I updated to resolve all code smells.  Added some time stamp logging to track execution times of the split and upload. 

Stuff I did not do:  More thorough unit tests and any kind of wrapper over this file splitter to make it a service.  It will likely be called directly from the api start job api in this state.

**Review by ncovercash — CHANGES_REQUESTED** [review]
I haven't tested the code properly, but here are some initial comments/questions!

File: src/main/java/org/folio/service/processing/split/FileSplitUtilities.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package org.folio.service.processing.split;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
[the remark this answers is not in the document]
Comment (ncovercash) [reply src/main/java/org/folio/service/processing/split/FileSplitUtilities.java]: Been getting this:
```
[WARNING] You aren't using a compiler supported by lombok, so lombok will not work and has been disabled.
Your processor is: org.aspectj.org.eclipse.jdt.internal.compiler.apt.dispatch.BatchProcessingEnvImpl
Lombok supports: OpenJDK javac, ECJ
        <unknown source file>:<no line information>
```

added a traditional constructor with exception

**Comment by ncovercash** [comment]
making a clean PR.

**Pull request closed without merging**